dev-tools/releng/src/main.rs

```diff
@@ -0,0 +1,669 @@
+                concat!($target_name, "-package"),
+                Command::new(&omicron_package).args([
+                    "--target",
+                    $target_name,
+                    "--artifacts",
+                    artifacts_path.as_str(),
+                    "package",
+                ]),
+            )
+            .after(concat!($target_name, "-target"));
+
+            jobs.push(
+                concat!($target_name, "-stamp"),
+                stamp_packages(
+                    logger.clone(),
+                    permits.clone(),
+                    args.output_dir.clone(),
+                    omicron_package.clone(),
+                    $target_name,
+                    artifacts_path.clone(),
+                    version.clone(),
+                    $proto_packages.iter().map(|(name, _)| *name),
+                ),
+            )
+            .after(concat!($target_name, "-package"));
+
+            let proto_dir = tempdir.path().join("proto").join($target_name);
+            jobs.push(
+                concat!($target_name, "-proto"),
+                build_proto_area(
+                    artifacts_path,
+                    proto_dir.clone(),
+                    &$proto_packages,
+                    manifest.clone(),
+                ),
+            )
+            .after(concat!($target_name, "-stamp"));
+
+            // The ${os_short_commit} token will be expanded by `helios-build`
+            let image_name = format!(
+                "{} {}/${{os_short_commit}} {}",
+                $image_prefix,
+                commit.chars().take(7).collect::<String>(),
+                Utc::now().format("%Y-%m-%d %H:%M")
+            );
+
+            jobs.push_command(
+                concat!($target_name, "-image"),
+                Command::new("ptime")
+                    .arg("-m")
+                    .arg(args.helios_dir.join("helios-build"))
+                    .arg("experiment-image")
+                    .arg("-o") // output directory for image
+                    .arg(args.output_dir.join(concat!("os-", $target_name)))
+                    .arg("-p") // use an external package repository
+                    .arg(format!("helios-dev={}", HELIOS_REPO))
+                    .arg("-F") // pass extra image builder features
+                    .arg(format!("optever={}", opte_version.trim()))
+                    .arg("-P") // include all files from extra proto area
+                    .arg(proto_dir.join("root"))
+                    .arg("-N") // image name
+                    .arg(image_name)
+                    .arg("-s") // tempdir name suffix
+                    .arg($target_name)
+                    .args($image_build_args)
+                    .current_dir(&args.helios_dir)
+                    .env("IMAGE_DATASET", &$image_dataset),
+            )
+            .after("helios-setup")
+            .after(concat!($target_name, "-proto"));
+        };
+    }
+
+    os_image_jobs! {
```

Comment:
Nice abstraction! Can this be a function call? If not, then can the reason for that be documented in a comment?

Reply:
I ended up refactoring this particular macro out but another macro slipped in. But I did comment why the new one isn't a function call.